File: implementation/servers/java/Channel.java
import java.util.ArrayList;
import java.util.stream.Stream;
import java.util.stream.Collectors;

class Channel
{
    // channel name
    private String name = "";

    // ctor
    public Channel(String name)
    {
        this.name = name;
    }

    public String getName()
    {
        // return channel name
        return this.name;
    }

    public void sendNicklist(User target)
    {
        ArrayList<User> users = Globals.server.getChannelUsers(this.name);

        String nicklist = users.stream().map(u -> u.getName()).collect(Collectors.joining(" "));

        target.sendMessage(353, target.getName() + " @ " + this.name, nicklist);
        target.sendMessage(366, target.getName() + " " + this.name, "END of /NAMES list.");
    }

    public void sendMessage(String source, String message)
    {
        Globals.server.sendChannelMessage(this.name, source, "PRIVMSG", this.name, message, false);
    }

    public void sendAll(String source, String type, String param, String message)
    {
        // send message to all channel members
        Globals.server.sendChannelMessage(this.name, source, type, param, message, true);
    }

    public void handleJoin(User user)
    {
        // send JOIN message to all channel members
        Globals.server.sendChannelMessage(this.name, user.getIdent(), "JOIN", this.name, "", true);

        // acknowledge join to user
        user.sendMessage(user.getIdent(), "JOIN", this.name, "");

        Globals.server.addChannelUser(this.name, user);

        sendNicklist(user);
    }

    public void handlePart(User user)
    {
        // send PART message to all channel members including sender 
        Globals.server.sendChannelMessage(this.name, user.getIdent(), "PART", this.name, "", true);
        Globals.server.removeChannelUser(this.name, user);
    }

    public void handleNick(User user, String nick)
    {
        // notify all channel members of nick change
        Globals.server.sendChannelMessage(this.name, user.getName(), "NICK", nick, "", true);
    }

    public void handleQuit(User user, String reason)
    {
        // tell all channel members that 'user' quitted 
        Globals.server.sendChannelMessage(this.name, user.getIdent(), "QUIT", "", reason, true);
        Globals.server.removeChannelUser(this.name, user);
    }
}

File: implementation/servers/cpp/Channel.cpp
#include "Channel.h"
#include <boost/algorithm/string.hpp>

Channel::Channel(const std::string &name): name{name} {

}

const std::string&Channel::get_name() const {
    return name;
}

const std::unordered_map<std::string, boost::shared_ptr<User>> Channel::get_users() const {
    return users;
}

void Channel::handle_join(boost::shared_ptr<User> user) {
    // Add the client to the internal user list
    users.insert({user -> get_name(), user});

    // Write all that the user joined
    send_all(user -> get_ident(), "JOIN", "");

    // Build a userlist
    std::string userlist;
    for(auto& u: users) {
        userlist += u.second -> get_name() + " ";
    }

    std::string m = user -> get_name() + " @ " + name;
    std::string n = user -> get_name() + " " + name;

    user -> send_message(353, m, userlist);
    user -> send_message(366, n, std::string("End of /NAMES list."));
}

void Channel::handle_part(boost::shared_ptr<User> user, const std::string& message) {
    send_all(user -> get_ident(), "PART", message);
    users.erase(user -> get_name());
}

void Channel::send_all(const std::string &source, const std::string &type, const std::string &message) {
    for (auto &it : users) {
        it.second -> send_message(source, type, name, message);
    }
}

void Channel::send_message(const std::string &source, const std::string &message) {
    // Send message to all but self
    for (auto &it : users) {
        auto &user = it.second;

        if(it.second -> get_ident() != source) {
            user -> send_message(source, "PRIVMSG", name, message);
        }
    }
}

void Channel::handle_nick(boost::shared_ptr<User> user, const std::string& new_nick) {
    for(auto &it: users) {
        it.second -> send_message(user -> get_ident(), "NICK", new_nick, "");
    }
}

void Channel::handle_quit(boost::shared_ptr<User> user, const std::string& message) {
    for(auto &it: users) {
        it.second -> send_message(user -> get_ident(), "QUIT", "", message);
    }

    users.erase(user -> get_name());
}
